File: Sources/RunestoneTextStorage/HighlightTextStorage.m
//
//  HighlightTextStorage.m
//  
//
//  Created by Simon Støvring on 29/11/2020.
//

#import "HighlightTextStorage.h"
@import RunestoneDocumentLineTree;
@import TreeSitterBindings;
@import TreeSitterJSON;

@interface HighlightTextStorage () <LineManagerDelegate>
@end

@implementation HighlightTextStorage {
    NSMutableAttributedString *_internalString;
    LineManager *_lineManager;
    Parser *_parser;
}

// MARK: - Lifecycle

- (instancetype)init {
    if (self = [super init]) {
        _internalString = [NSMutableAttributedString new];
        _lineManager = [LineManager new];
        _lineManager.delegate = self;
        _parser = [Parser new];
        _parser.language = [[Language alloc] initWithLanguage:tree_sitter_json()];
    }
    return self;
}

// MARK: - NSTextStorage

- (NSString *)string {
    return _internalString.string;
}

- (void)replaceCharactersInRange:(NSRange)range withString:(NSString *)str {
    [self beginEditing];
    NSInteger length = (NSInteger)str.length - (NSInteger)range.length;
    [_internalString replaceCharactersInRange:range withString:str];
    [_lineManager removeCharactersInRange:range];
    [_lineManager insertString:str inRange:range];
    [self edited:NSTextStorageEditedCharacters range:range changeInLength:length];
    [self endEditing];
}

- (void)setAttributes:(NSDictionary<NSAttributedStringKey,id> *)attrs range:(NSRange)range {
    [self beginEditing];
    [_internalString setAttributes:attrs range:range];
    [self edited:NSTextStorageEditedAttributes range:range changeInLength:0];
    [self endEditing];
}

- (NSDictionary<NSAttributedStringKey,id> *)attributesAtIndex:(NSUInteger)location effectiveRange:(NSRangePointer)range {
    return [_internalString attributesAtIndex:location effectiveRange:range];
}

- (void)processEditing {
    [super processEditing];
//    [_parser parse:self.string];
}

// MARK: - HighlightTextStorage

- (NSString * _Nonnull)lineManager:(LineManager * _Nonnull)lineManager characterAtLocation:(NSInteger)location {
    return [self.string substringWithRange:NSMakeRange(location, 1)];
}

@end

File: Sources/Runestone/UI/HighlightTextStorage.swift
//
//  HighlightTextStorage.swift
//  
//
//  Created by Simon Støvring on 29/11/2020.
//

import UIKit
import TreeSitterBindings
import TreeSitterJSON

final class HighlightTextStorage: NSTextStorage {
    override var string: String {
        return internalString.string
    }

    private let internalString = NSMutableAttributedString()

    override init() {
        super.init()
        tree_sitter_json()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }

    override func replaceCharacters(in range: NSRange, with str: String) {
        beginEditing()
        internalString.replaceCharacters(in: range, with: str)
        edited(.editedCharacters, range: range, changeInLength: str.count - range.length)
        endEditing()
    }

    override func setAttributes(_ attrs: [NSAttributedString.Key : Any]?, range: NSRange) {
        beginEditing()
        internalString.setAttributes(attrs, range: range)
        edited(.editedAttributes, range: range, changeInLength: 0)
        endEditing()
    }

    override func attributes(at location: Int, effectiveRange range: NSRangePointer?) -> [NSAttributedString.Key: Any] {
        return internalString.attributes(at: location, effectiveRange: range)
    }

    override func processEditing() {
        super.processEditing()
//        let editedRange = string.convert(self.editedRange)
//        let lineRange = self.lineRange(from: editedRange)
//        let lineNSRange = string.convert(lineRange)
//        internalString.removeAttribute(.foregroundColor, range: lineNSRange)
//        if let tokenizer = tokenizer {
//            let tokens = tokenizer.tokenize(String(string[lineRange]))
//            for token in tokens {
//                let rangeInLine = token.range
//                let rangeInText = NSMakeRange(lineNSRange.location + rangeInLine.location, rangeInLine.length)
//                internalString.addAttributes([.foregroundColor: UIColor.red], range: rangeInText)
//            }
//        }
    }
}

private extension HighlightTextStorage {
    private func lineRange(from range: Range<String.Index>) -> Range<String.Index> {
        var lineStart: String.Index = range.lowerBound
        var lineEnd: String.Index = range.upperBound
        var contentsEnd: String.Index = range.upperBound
        string.getLineStart(&lineStart, end: &lineEnd, contentsEnd: &contentsEnd, for: range)
        return lineStart ..< lineEnd
    }
}
